File: tests/machine/x/pascal/outer_join.pas
program OuterJoin;
{$mode objfpc}
{$modeswitch nestedprocvars}

uses SysUtils, fgl, fphttpclient, Classes, Variants, fpjson, jsonparser, fpjsonrtti;

type
  generic TArray<T> = array of T;

var
  _tmp0: specialize TFPGMap<string, Variant>;
  _tmp1: specialize TFPGMap<string, Variant>;
  _tmp2: specialize TFPGMap<string, Variant>;
  _tmp3: specialize TFPGMap<string, Variant>;
  _tmp4: specialize TFPGMap<string, integer>;
  _tmp5: specialize TFPGMap<string, integer>;
  _tmp6: specialize TFPGMap<string, integer>;
  _tmp7: specialize TFPGMap<string, integer>;
  _tmp8: specialize TArray<specialize TFPGMap<string, Variant>>;
  _tmp9: specialize TFPGMap<string, Variant>;
  customers: specialize TArray<specialize TFPGMap<string, Variant>>;
  o: specialize TFPGMap<string, integer>;
  orders: specialize TArray<specialize TFPGMap<string, integer>>;
  _result: specialize TArray<specialize TFPGMap<string, Variant>>;
  row: specialize TFPGMap<string, Variant>;

begin
  _tmp0 := specialize TFPGMap<string, Variant>.Create;
  _tmp0.AddOrSetData('id', 1);
  _tmp0.AddOrSetData('name', 'Alice');
  _tmp1 := specialize TFPGMap<string, Variant>.Create;
  _tmp1.AddOrSetData('id', 2);
  _tmp1.AddOrSetData('name', 'Bob');
  _tmp2 := specialize TFPGMap<string, Variant>.Create;
  _tmp2.AddOrSetData('id', 3);
  _tmp2.AddOrSetData('name', 'Charlie');
  _tmp3 := specialize TFPGMap<string, Variant>.Create;
  _tmp3.AddOrSetData('id', 4);
  _tmp3.AddOrSetData('name', 'Diana');
  customers := specialize TArray<specialize TFPGMap<string, Variant>>([_tmp0, _tmp1, _tmp2, _tmp3]);
  _tmp4 := specialize TFPGMap<string, integer>.Create;
  _tmp4.AddOrSetData('id', 100);
  _tmp4.AddOrSetData('customerId', 1);
  _tmp4.AddOrSetData('total', 250);
  _tmp5 := specialize TFPGMap<string, integer>.Create;
  _tmp5.AddOrSetData('id', 101);
  _tmp5.AddOrSetData('customerId', 2);
  _tmp5.AddOrSetData('total', 125);
  _tmp6 := specialize TFPGMap<string, integer>.Create;
  _tmp6.AddOrSetData('id', 102);
  _tmp6.AddOrSetData('customerId', 1);
  _tmp6.AddOrSetData('total', 300);
  _tmp7 := specialize TFPGMap<string, integer>.Create;
  _tmp7.AddOrSetData('id', 103);
  _tmp7.AddOrSetData('customerId', 5);
  _tmp7.AddOrSetData('total', 80);
  orders := specialize TArray<specialize TFPGMap<string, integer>>([_tmp4, _tmp5, _tmp6, _tmp7]);
  SetLength(_tmp8, 0);
  for o in orders do
    begin
      for c in customers do
        begin
          if not ((o.KeyData['customerId'] = c.id)) then continue;
          _tmp9 := specialize TFPGMap<string, Variant>.Create;
          _tmp9.AddOrSetData('order', o);
          _tmp9.AddOrSetData('customer', c);
          _tmp8 := Concat(_tmp8, [_tmp9]);
        end;
    end;
  _result := _tmp8;
  writeln('--- Outer Join using syntax ---');
  for row in _result do
    begin
      if row.KeyData['order'] then
        begin
          if row.KeyData['customer'] then
            begin
              writeln('Order', ' ', row.KeyData['order'].id, ' ', 'by', ' ', row.KeyData['customer']
                      .name, ' ', '- $', ' ', row.KeyData['order'].total);
            end
          else
            begin
              writeln('Order', ' ', row.KeyData['order'].id, ' ', 'by', ' ', 'Unknown', ' ', '- $',
                      ' ', row.KeyData['order'].total);
            end;
        end
      else
        begin
          writeln('Customer', ' ', row.KeyData['customer'].name, ' ', 'has no orders');
        end;
    end;
end.

File: tests/machine/x/scala/outer_join.scala
object outer_join {
  case class Customer(id: Int, name: String)
  case class Order(id: Int, customerId: Int, total: Int)
  case class Result(order: Option[Order], customer: Option[Customer])
  case class Result1(order: Order, customer: Customer)

  def _outer_join[A,B](a: List[A], b: List[B])(cond: (A,B) => Boolean): List[(Option[A], Option[B])] = { val left = _left_join(a,b)(cond).map{ case(x,y) => (Some(x), y) }; val right = _right_join(a,b)(cond).collect{ case(None, r) => (None, Some(r)) }; left ++ right }

  val customers = List[Customer](Customer(id = 1, name = "Alice"), Customer(id = 2, name = "Bob"), Customer(id = 3, name = "Charlie"), Customer(id = 4, name = "Diana"))
  val orders = List[Order](Order(id = 100, customerId = 1, total = 250), Order(id = 101, customerId = 2, total = 125), Order(id = 102, customerId = 1, total = 300), Order(id = 103, customerId = 5, total = 80))
  val result = for { _pair0 <- _outer_join(orders, customers)((o,c) => o.customerId == (c.id).asInstanceOf[Int]); o = _pair0._1; c = _pair0._2 } yield Result(order = o, customer = c)
  def main(args: Array[String]): Unit = {
    println(("--- Outer Join using syntax ---"))
    for(row <- result) {
      if (row.order != null) {
        if (row.customer != null) {
          println(("Order") + " " + (row.order.id) + " " + ("by") + " " + (row.customer.name) + " " + ("- $") + " " + (row.order.total))
        } else {
          println(("Order") + " " + (row.order.id) + " " + ("by") + " " + ("Unknown") + " " + ("- $") + " " + (row.order.total))
        }
      } else {
        println(("Customer") + " " + (row.customer.name) + " " + ("has no orders"))
      }
    }
  }
}
